zerver/views/auth.py

```diff
@@ -855,6 +855,61 @@ def view(request, *args, **kwargs):
     return two_fa_view(request, **kwargs)
 
 
+@csrf_exempt
+@has_request_variables
+def jwt_fetch_api_key(request: HttpRequest, json_web_token: str = REQ(default="")) -> HttpResponse:
+    realm = get_realm_from_request(request)
+    if realm is None:
+        raise JsonableError(_("Invalid subdomain"))
+
+    host = realm.host
+    try:
+        key = settings.JWT_FETCH_API_KEYS[host]["key"]
+        algorithms = settings.JWT_FETCH_API_KEYS[host]["algorithms"]
+    except KeyError:
+        raise JsonableError(_("Auth key for this subdomain not found"))
+
+    try:
+        json_web_token = request.POST["json_web_token"]
+        options = {"verify_signature": True}
+        payload = jwt.decode(json_web_token, key, algorithms=algorithms, options=options)
+    except KeyError:
+        raise JsonableError(_("No JSON web token passed in request"))
+    except jwt.InvalidTokenError:
+        raise JsonableError(_("Bad JSON web token"))
+
+    remote_email = payload.get("email", None)
+    if remote_email is None or not remote_email:
+        raise JsonableError(_("No email specified in JSON web token claims"))
+
+    return_data: Dict[str, bool] = {}
+    user_profile = authenticate(
+        username=remote_email, realm=realm, return_data=return_data, use_dummy_backend=True
+    )
+    if return_data.get("inactive_user"):
+        raise JsonableError(_("Account is deactivated"))
+    if return_data.get("inactive_realm"):
+        raise JsonableError(_("This organization has been deactivated"))
+    if return_data.get("invalid_subdomain"):
+        raise JsonableError(_("Invalid subdomain"))
+    if user_profile is None:
+        raise JsonableError(_("User not found"))
+
+    process_client(request, user_profile)
+
+    api_key = get_api_key(user_profile)
+    members = get_raw_user_data(
+        realm,
+        user_profile,
+        target_user=user_profile,
+        client_gravatar=False,
+        user_avatar_url_field_optional=False,
+        include_custom_profile_fields=False,
+    )
```

Comment:
The ~20 lines ending here have a lot of overlap with, say, `api_fetch_api_key`.

And I think where they differ, the other implementation is better. Can we extract this block:

```
    if return_data.get("inactive_user"):                                                         
        raise UserDeactivatedError()                                                             
    if return_data.get("inactive_realm"):                                                        
        raise RealmDeactivatedError()                                                            
    if return_data.get("password_auth_disabled"):                                                
        raise PasswordAuthDisabledError()                                                        
    if return_data.get("password_reset_needed"):                                                 
        raise PasswordResetRequiredError()                                                       
    if user_profile is None:                                                                     
        raise AuthenticationFailedError()                                                        
                                                                                                 
    assert user_profile.is_authenticated                                                         
                                                                                                 
    # Maybe sending 'user_logged_in' signal is the better approach:                              
    #   user_logged_in.send(sender=user_profile.__class__, request=request, user=user_profile)   
    # Not doing this only because over here we don't add the user information                    
    # in the session. If the signal receiver assumes that we do then that                        
    # would cause problems.                                                                      
    email_on_new_login(sender=user_profile.__class__, request=request, user=user_profile)        
                                                                                                 
    # Mark this request as having a logged-in user for our server logs.                          
    assert isinstance(user_profile, UserProfile)                                                 
    process_client(request, user_profile)                                                        
    RequestNotes.get_notes(request).requestor_for_logs = user_profile.format_requestor_for_logs()
                                                                                                 
    api_key = get_api_key(user_profile)                                                          
    return json_success(request, data={"api_key": api_key, "email": user_profile.delivery_email})
```

into a shared function, perhaps with an interface `return process_api_key_fetch_authenticate_result(request, user_profile, return_data)`?

Reply:
Yes there is some overlap between the functions.

1) One key difference is that our `jwt_fetch_api_key` is authenticating the user through the dummy backend.

So I guess a shared helper function should make optional some steps, like
```python
email_on_new_login(sender=user_profile.__class__, request=request, user=user_profile)
```
that is something to definitely skip in our case.

Maybe also these calls are totally pointless for `jwt_fetch_api_key`?
```python
assert user_profile.is_authenticated
process_client(request, user_profile)
RequestNotes.get_notes(request).requestor_for_logs = user_profile.format_requestor_for_logs()
```
We already have a call to `process_client` but I suspect that is useless for our purpose.

2) Should we decide to implement a common helper, which source file should be modified? I don't think `auth.py` is the right one here.

3) Meanwhile inspecting the `api_fetch_api_key` I've noticed that we can re-use the specific error classes instead of creating Jsonable Errors on the fly, I'll do this asap.

Reply:
Can you explain why skipping `email_on_new_login` is correct for your case? I would expect an app integrated with Zulip to only need to do JWT authentication once on setup, and then use the API key thereafter, and I think it'd be correct to email users that this login process has happened. To make the emails clear, we would need to add the bit of plumbing to make sure the device/client details in those emails are clear (i.e. that it's a logic to the relevant third-party app).

The `process_client` call is not incorrect/useless; it does logging that's of modest value.

Reply:
At IETF meetings, each meeting session can be seen as a sort of standalone app which uses Zulip as the chat backend. Every time you join a meeting session, we use the JWT authentication to retrieve your apikey ans use it for direct/private messaging. JWT might also be performed more than once during a session, e.g., if a user changes their apikey from the Zulip interface.

I feel that sending emails every time you join a session (which could happen quite some times during the IETF meeting week, even simultaneously) would be bothersome from a user perspective and maybe also add unnecessary load to the Zulip server? Pinging @rjsparks as he may have a take on this.

Reply:
It would be extremely bothersome to emit email for each meeting session joined. An a new oidc handshake is necessary for each new session as details about the user returned in the oidc claims may change between (or even during, requiring a logout, login cycle) meeting sessions.

Reply:
@rjsparks There's the `SEND_LOGIN_EMAILS` setting in `/etc/zulip/settings.py` that you can use to disable these email, so I think that should ease this concern for your use case?
```
## Controls whether Zulip sends "new login" email notifications.
# SEND_LOGIN_EMAILS = True
```

Reply:
@mateuszmandera : Well, that would also turn off the mails in the more mundane use case where people actually logged in from a new client that wasn't this particular one. Having the granularity to affect the behavior depending on how the connection is being made seems like it would be a good thing? But I will discuss using that setting to disable the new login emails sitewide with the team here.

Reply:
> Stepping back a bit, from a security model perspective, Zulip should be notifying users every time it's handing out an authentication cookie or API key to some external source

Fair enough. Let's leave the email sending there. We may then globally disable email sending via `SEND_LOGIN_EMAILS` or work on caching apikeys.

Reply:
@timabbott fyi I'm ready to update the PR and add the helper function refactoring too

Reply:
The PR has been updated, the code now includes a helper function 
```python
def process_api_key_fetch_authenticate_result(
   request: HttpRequest, 
   user_profile: UserProfile, 
   return_data: Dict[str, bool]
) -> str:
```

that is used by `jwt_fetch_api_key` and `api_fetch_api_key`